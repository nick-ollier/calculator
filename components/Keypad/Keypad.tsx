import classnames from 'classnames';
import { useCalculator } from '@hooks';
import Button from '@components/Button/Button';
import { keypadExtended, keypadDefault } from '@constants';

export interface ExtendedProps {
    portrait: boolean;
}

const Extended: React.FC<ExtendedProps> = ({ portrait }) => {
    const {
        state: { activeOperator },
        handleNumber,
        handleClear,
        handlePlusMinus,
        handlePercent,
        handleOperator,
        handleScientific,
        handleEquals
    } = useCalculator(portrait);

    const keypad = portrait ? keypadDefault : keypadExtended;

    return (
        <div
            className={classnames('grid', {
                'grid-cols-4 gap-2 mb-6 px-1 text-2xl': portrait,
                'grid-cols-10 gap-1 px-4': !portrait
            })}
            data-testid='keypad'
        >
            {keypad.map(({ variant, value, span, label, accessibleLabel }) => (
                <Button
                    key={label}
                    variant={variant}
                    value={value}
                    span={span}
                    label={label}
                    accessibleLabel={accessibleLabel || label}
                    onClick={() => {
                        if (variant === 'operator' && value !== '=')
                            handleOperator(value);
                        if (variant === 'number') handleNumber(label);
                        /* istanbul ignore next */
                        if (variant === 'scientific') handleScientific();
                        if (value === '=') handleEquals();
                        if (value === 'clear') handleClear();
                        if (value === 'plusMinus') handlePlusMinus();
                        if (value === 'percent') handlePercent();
                    }}
                    active={activeOperator === value}
                    portrait={portrait}
                />
            ))}
        </div>
    );
};

export default Extended;
