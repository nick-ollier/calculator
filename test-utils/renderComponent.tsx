import CalculatorProvider from '@context/CalculatorContext';
import { render } from '@testing-library/react';

const renderWrapper = (ui: React.ReactElement) => {
    const WithProviders = ({ children }) => {
        return <CalculatorProvider>{children}</CalculatorProvider>;
    };

    return render(ui, { wrapper: WithProviders });
};

export * from '@testing-library/react';
export { renderWrapper as render };
