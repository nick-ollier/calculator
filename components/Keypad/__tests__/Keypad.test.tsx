import { render, screen } from '@testUtils/renderComponent';
import userEvent from '@testing-library/user-event';
import { keypadExtended, keypadDefault } from '@constants';
import Keypad, { ExtendedProps } from '@components/Keypad/Keypad';

const renderComponent = (props?: Partial<ExtendedProps>) => {
    const utils = render(<Keypad portrait {...props} />);

    return {
        ...utils,
        keypad: screen.getByTestId('keypad')
    };
};

const mockHandleNumber = jest.fn();
const mockHandleClear = jest.fn();
const mockHandlePlusMinus = jest.fn();
const mockHandlePercent = jest.fn();
const mockHandleOperator = jest.fn();
const mockHandleBackspace = jest.fn();
const mockHandleEquals = jest.fn();

jest.mock('@hooks/useCalculator', () => ({
    useCalculator: () => ({
        state: {
            displayValue: '',
            activeOperator: null,
            hasDecimal: false,
            expression: '',
            lastAction: '',
            hasChanged: false
        },
        handleNumber: mockHandleNumber,
        handleClear: mockHandleClear,
        handlePlusMinus: mockHandlePlusMinus,
        handlePercent: mockHandlePercent,
        handleOperator: mockHandleOperator,
        handleBackspace: mockHandleBackspace,
        handleEquals: mockHandleEquals
    })
}));

describe('component: keypad', () => {
    describe('portrait', () => {
        describe('default', () => {
            it('renders', () => {
                const { container } = renderComponent();
                expect(container);
            });

            it('renders with correct classes', () => {
                const { keypad } = renderComponent();
                expect(keypad).toHaveClass(
                    'grid-cols-4 gap-2 mb-6 px-1 text-2xl'
                );
            });

            it('renders correct number of keypad buttons', () => {
                renderComponent();
                expect(screen.getAllByRole('button').length).toEqual(
                    keypadDefault.length
                );
            });

            it('matches snapshot', () => {
                const { asFragment } = renderComponent();
                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe('button callbacks', () => {
            it('handleNumber', () => {
                renderComponent();
                const button = screen.getByRole('button', {
                    name: /1/i
                });

                userEvent.click(button);

                expect(mockHandleNumber).toHaveBeenCalledWith('1');
                expect(mockHandleNumber).toHaveBeenCalledTimes(1);
            });

            it('handleClear', () => {
                renderComponent();
                const button = screen.getByRole('button', {
                    name: /clear/i
                });

                userEvent.click(button);

                expect(mockHandleClear).toHaveBeenCalledWith();
                expect(mockHandleClear).toHaveBeenCalledTimes(1);
            });

            it('handlePlusMinus', () => {
                renderComponent();
                const button = screen.getByRole('button', {
                    name: /plus \/ minus/i
                });

                userEvent.click(button);

                expect(mockHandlePlusMinus).toHaveBeenCalledWith();
                expect(mockHandlePlusMinus).toHaveBeenCalledTimes(1);
            });

            it('handlePercent', () => {
                renderComponent();
                const button = screen.getByRole('button', {
                    name: /percent/i
                });

                userEvent.click(button);

                expect(mockHandlePercent).toHaveBeenCalledWith();
                expect(mockHandlePercent).toHaveBeenCalledTimes(1);
            });

            it('handleOperator', () => {
                renderComponent();
                const button = screen.getByRole('button', {
                    name: /multiply/i
                });

                userEvent.click(button);

                expect(mockHandleOperator).toHaveBeenCalledWith('*');
                expect(mockHandleOperator).toHaveBeenCalledTimes(1);
            });

            it('handleEquals', () => {
                renderComponent();
                const button = screen.getByRole('button', {
                    name: /equals/i
                });

                userEvent.click(button);

                expect(mockHandleEquals).toHaveBeenCalledWith();
                expect(mockHandleEquals).toHaveBeenCalledTimes(1);
            });
        });
    });

    describe('landscape', () => {
        describe('default', () => {
            it('renders', () => {
                const { container } = renderComponent({ portrait: false });
                expect(container);
            });

            it('renders with correct classes', () => {
                const { keypad } = renderComponent({ portrait: false });
                expect(keypad).toHaveClass('grid-cols-10 gap-1 px-4');
            });

            it('renders correct number of keypad buttons', () => {
                renderComponent({ portrait: false });
                expect(screen.getAllByRole('button').length).toEqual(
                    keypadExtended.length
                );
            });

            it('matches snapshot', () => {
                const { asFragment } = renderComponent({ portrait: false });
                expect(asFragment()).toMatchSnapshot();
            });
        });
    });
});
