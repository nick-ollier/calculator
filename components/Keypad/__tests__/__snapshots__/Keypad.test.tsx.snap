// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: keypad landscape default matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-10 gap-1 px-4"
    data-testid="keypad"
  >
    <button
      aria-label="("
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      (
    </button>
    <button
      aria-label=")"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      )
    </button>
    <button
      aria-label="mc"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      mc
    </button>
    <button
      aria-label="m+"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      m+
    </button>
    <button
      aria-label="m-"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      m-
    </button>
    <button
      aria-label="mr"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      mr
    </button>
    <button
      aria-label="Clear"
      class="rounded-full col-span-1 bg-gray-200 text-gray-500 min-h-btn-l min-w-btn-l"
      type="button"
      value="clear"
    >
      AC
    </button>
    <button
      aria-label="Plus / Minus"
      class="rounded-full col-span-1 bg-gray-200 text-gray-500 min-h-btn-l min-w-btn-l"
      type="button"
      value="plusMinus"
    >
      <sup>
        +
      </sup>
      /
      <sub>
        −
      </sub>
    </button>
    <button
      aria-label="Percent"
      class="rounded-full col-span-1 bg-gray-200 text-gray-500 min-h-btn-l min-w-btn-l"
      type="button"
      value="percent"
    >
      %
    </button>
    <button
      aria-label="Divide"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="/"
    >
      ÷
    </button>
    <button
      aria-label="2<sup>nd</sup>"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      2
      <sup>
        nd
      </sup>
    </button>
    <button
      aria-label="x&sup2;"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      x²
    </button>
    <button
      aria-label="x&sup3;"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      x³
    </button>
    <button
      aria-label="x<sup>y</sup>"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      x
      <sup>
        y
      </sup>
    </button>
    <button
      aria-label="e<sup>x</sup>"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      e
      <sup>
        x
      </sup>
    </button>
    <button
      aria-label="10<sup>x</sup>"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      10
      <sup>
        x
      </sup>
    </button>
    <button
      aria-label="7"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="7"
    >
      7
    </button>
    <button
      aria-label="8"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="8"
    >
      8
    </button>
    <button
      aria-label="9"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="9"
    >
      9
    </button>
    <button
      aria-label="Multiply"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="*"
    >
      ×
    </button>
    <button
      aria-label="<sup>1</sup>/<sub>x</sub>"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      <sup>
        1
      </sup>
      /
      <sub>
        x
      </sub>
    </button>
    <button
      aria-label="&sup2;&radic;x"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      ²√x
    </button>
    <button
      aria-label="&sup3;&radic;x"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      ³√x
    </button>
    <button
      aria-label="<sup>y</sup>&radic;x"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      <sup>
        y
      </sup>
      √x
    </button>
    <button
      aria-label="In"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      In
    </button>
    <button
      aria-label="log&#8321;&#8320;"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      log₁₀
    </button>
    <button
      aria-label="4"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="4"
    >
      4
    </button>
    <button
      aria-label="5"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="5"
    >
      5
    </button>
    <button
      aria-label="6"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="6"
    >
      6
    </button>
    <button
      aria-label="Minus"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="-"
    >
      −
    </button>
    <button
      aria-label="x!"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      x!
    </button>
    <button
      aria-label="sin"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      sin
    </button>
    <button
      aria-label="cos"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      cos
    </button>
    <button
      aria-label="tan"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      tan
    </button>
    <button
      aria-label="e"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      e
    </button>
    <button
      aria-label="EE"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      EE
    </button>
    <button
      aria-label="1"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="1"
    >
      1
    </button>
    <button
      aria-label="2"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="2"
    >
      2
    </button>
    <button
      aria-label="3"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="3"
    >
      3
    </button>
    <button
      aria-label="Plus"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="+"
    >
      +
    </button>
    <button
      aria-label="Rad"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      Rad
    </button>
    <button
      aria-label="sinh"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      sinh
    </button>
    <button
      aria-label="cosh"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      cosh
    </button>
    <button
      aria-label="tanh"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      tanh
    </button>
    <button
      aria-label="&pi;"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      π
    </button>
    <button
      aria-label="Rand"
      class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
    >
      Rand
    </button>
    <button
      aria-label="0"
      class="rounded-full col-span-2 bg-gray-400 text-gray-100 text-left px-6 min-h-btn-l min-w-btn-l"
      type="button"
      value="0"
    >
      0
    </button>
    <button
      aria-label="."
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="."
    >
      .
    </button>
    <button
      aria-label="Equals"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-h-btn-l min-w-btn-l"
      type="button"
      value="="
    >
      =
    </button>
  </div>
</DocumentFragment>
`;

exports[`component: keypad portrait default matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-4 gap-2 mb-6 px-1 text-2xl"
    data-testid="keypad"
  >
    <button
      aria-label="Clear"
      class="rounded-full col-span-1 bg-gray-200 text-gray-500 min-w-btn-p min-h-btn-p"
      type="button"
      value="clear"
    >
      AC
    </button>
    <button
      aria-label="Plus / Minus"
      class="rounded-full col-span-1 bg-gray-200 text-gray-500 min-w-btn-p min-h-btn-p"
      type="button"
      value="plusMinus"
    >
      <sup>
        +
      </sup>
      /
      <sub>
        −
      </sub>
    </button>
    <button
      aria-label="Percent"
      class="rounded-full col-span-1 bg-gray-200 text-gray-500 min-w-btn-p min-h-btn-p"
      type="button"
      value="percent"
    >
      %
    </button>
    <button
      aria-label="Divide"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="/"
    >
      ÷
    </button>
    <button
      aria-label="7"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="7"
    >
      7
    </button>
    <button
      aria-label="8"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="8"
    >
      8
    </button>
    <button
      aria-label="9"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="9"
    >
      9
    </button>
    <button
      aria-label="Multiply"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="*"
    >
      ×
    </button>
    <button
      aria-label="4"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="4"
    >
      4
    </button>
    <button
      aria-label="5"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="5"
    >
      5
    </button>
    <button
      aria-label="6"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="6"
    >
      6
    </button>
    <button
      aria-label="Minus"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="-"
    >
      −
    </button>
    <button
      aria-label="1"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="1"
    >
      1
    </button>
    <button
      aria-label="2"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="2"
    >
      2
    </button>
    <button
      aria-label="3"
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="3"
    >
      3
    </button>
    <button
      aria-label="Plus"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="+"
    >
      +
    </button>
    <button
      aria-label="0"
      class="rounded-full col-span-2 bg-gray-400 text-gray-100 text-left px-6 min-w-btn-p min-h-btn-p"
      type="button"
      value="0"
    >
      0
    </button>
    <button
      aria-label="."
      class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="."
    >
      .
    </button>
    <button
      aria-label="Equals"
      class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-w-btn-p min-h-btn-p"
      type="button"
      value="="
    >
      =
    </button>
  </div>
</DocumentFragment>
`;
